### PR TITLE
[Wsign-compare] warning resolutions, by directory, 2

### PR DESCRIPTION
**Pull request by tg-at-google**
Resolution affecting directory:
`tensorflow/compiler/mlir/lite`

@mihaimaruseac

### CODE REVIEW

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc [hunk 1]
```diff
@@ -147,18 +147,10 @@ bool IsI64Type(Type element_type) {
...
-      IsQUI8Type(element_type)) {
-    return VerifyOperandsHaveSameShapesOrBroadcastableShape(
-        /*op=*/op.getOperation(), /*indices=*/ArrayRef<unsigned>{0, 1},
-        /*max_bcast_rank=*/5);
-  }
-
-  // Allows I32 output when the operands have valid shapes, which are
-  // broadcastable shapes up to four dimension or have same shapes.
-  if (IsI32Type(element_type)) {
+      IsQUI8Type(element_type) || IsI32Type(element_type)) {
```
Comment (mihaimaruseac) [inline tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:153]: These look like changes outside of scope for the PR

File: tensorflow/compiler/mlir/lite/transforms/legalize_tf.cc [hunk 2]
```diff
@@ -170,7 +170,7 @@ LogicalResult ConvertTFRandomUniformOp::matchAndRewrite(
       size_t num_samples = Distribution::kResultElementCount;
       llvm::SmallVector<float, 32> data;
       data.resize(num_elements);
-      while (offset < num_elements) {
+      while (static_cast<int>(offset) < num_elements) {
```
Comment (mihaimaruseac) [inline tensorflow/compiler/mlir/lite/transforms/legalize_tf.cc:173]: I'd rather change type of `num_elements` local

File: tensorflow/compiler/mlir/lite/transforms/legalize_tf.cc [hunk 3]
```diff
@@ -631,6 +631,156 @@ struct LegalizeUnidirectionalSequenceRnn : public RewritePattern {
...
+  // Look if there is a function that tries until it converge.
+  // Currently unit-test doesn't do multiple tries, so we need this.
+  const int max_iterations = 15;
+  for (int i = 0; i < max_iterations; ++i) {
+    if (failed(applyPartialConversion(func, target, patterns))) {
+      return;
+    }
+  }
+}
+
```
Comment (mihaimaruseac) [inline tensorflow/compiler/mlir/lite/transforms/legalize_tf.cc:783]: This looks unrelated?

File: tensorflow/compiler/mlir/lite/transforms/legalize_tf.cc [hunk 4]
```diff
@@ -681,16 +831,32 @@ void LegalizeTF::runOnFunction() {
...
+                  ApplyExplicitBroadcasting<TF::SubOp>,
+                  ApplyExplicitBroadcasting<TF::FloorDivOp>,
+                  ApplyExplicitBroadcasting<TF::FloorModOp>,
+                  ApplyExplicitBroadcasting<TF::PowOp>,
+                  ApplyExplicitBroadcasting<TF::MaximumOp>,
+                  ApplyExplicitBroadcasting<TF::MinimumOp>,
+                  ApplyExplicitBroadcasting<TF::SquaredDifferenceOp>,
+                  ApplyExplicitBroadcasting<TF::SelectV2Op>>(context);
+
+  applyPatterns(func, target, patterns);
```
Comment (mihaimaruseac) [inline tensorflow/compiler/mlir/lite/transforms/legalize_tf.cc:859]: Unrelated change?

File: tensorflow/compiler/mlir/lite/transforms/optimize.cc [hunk 5]
```diff
@@ -198,7 +198,7 @@ DenseElementsAttr GetShape(Value output_val) {
   auto output_type = output_val.getType().cast<RankedTensorType>();
   auto shape_vector = output_type.getShape();
   std::vector<int32_t> shape(shape_vector.size());
-  for (int i = 0; i < shape_vector.size(); ++i) {
+  for (int i = 0, end = shape_vector.size(); i < end; ++i) {
     shape[i] = shape_vector[i];
   }
```
Comment (mihaimaruseac) [inline tensorflow/compiler/mlir/lite/transforms/optimize.cc:203]: This loop should be written using range loops and `emplace_back`/`push_back`.

File: tensorflow/compiler/mlir/lite/transforms/prepare_composite_functions_tf.cc [hunk 6]
```diff
@@ -225,7 +225,8 @@ void PrepareCompositeFunctionsPass::ConvertTFImplementsWithAttributes(
 LogicalResult CheckOutputConsumer(
     Operation* call_op, int expected_num_outputs,
     llvm::DenseSet<int> expected_consumer_indices) {
-  if (call_op->getNumResults() != expected_num_outputs) return failure();
+  const int call_op_getNumResults = call_op->getNumResults();
+  if (call_op_getNumResults != expected_num_outputs) return failure();
```
Comment (mihaimaruseac) [inline tensorflow/compiler/mlir/lite/transforms/prepare_composite_functions_tf.cc:229]:
```suggestion
  const int num_results = call_op->getNumResults();
  if (num_results != expected_num_outputs) return failure();
```

Much more readable name and follows style guide.

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
Please make sure to not include other unrelated changes

File: tensorflow/compiler/mlir/lite/transforms/optimize.cc (hunk 5 above)
Comment (tg-at-google) [reply tensorflow/compiler/mlir/lite/transforms/optimize.cc:203]: updated.

File: tensorflow/compiler/mlir/lite/transforms/legalize_tf.cc (hunk 4 above)
Comment (tg-at-google) [reply tensorflow/compiler/mlir/lite/transforms/legalize_tf.cc]: removed.

**Comment by tg-at-google** [comment]
all recommended updates applied.

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc [hunk 7]
```diff
@@ -773,7 +773,8 @@ static LogicalResult Verify(CustomOp op) {
       op.custom_option().cast<OpaqueElementsAttr>();
   if (!opaque_attr.getType().hasStaticShape())
     return op.emitOpError("custom_option should have a static shape.");
-  if (opaque_attr.getValue().size() !=
+  const int opaque_attr_getValue_size = opaque_attr.getValue().size();
+  if (opaque_attr_getValue_size !=
```
Comment (mihaimaruseac) [inline tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:777]: Please use readable name which also follows coding style

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc [hunk 8]
```diff
@@ -955,7 +956,7 @@ static LogicalResult Verify(ScatterNdOp op) {
     // Checks whether the last `(shape_type.getDimSize(0) - outermost_dim)`
     // dimensions of `updates` and `shape` are equal.
     for (auto shape_it : llvm::enumerate(shape_value)) {
-      auto i = shape_it.index();
+      long int i = shape_it.index();
```
Comment (mihaimaruseac) [inline tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:959]: Unsure about this change. What is the warning here?

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc [hunk 9]
```diff
@@ -1192,7 +1193,8 @@ struct RemoveRedundantUnpackPack : public RewritePattern {
       return failure();
 
     const int total_pack_inputs = pack_op.getNumOperands();
-    if (total_pack_inputs != input_unpack_op.getNumResults()) return failure();
+    const int input_unpack_op_getNumResults = input_unpack_op.getNumResults();
+    if (total_pack_inputs != input_unpack_op_getNumResults) return failure();
```
Comment (mihaimaruseac) [inline tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:1197]:
```suggestion
    const int num_results = input_unpack_op.getNumResults();
    if (total_pack_inputs != num_results) return failure();
```

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc [hunk 10]
```diff
@@ -1261,7 +1263,7 @@ static LogicalResult Verify(SliceOp op) {
   }
 
   if (begin && size && input_type.hasStaticShape()) {
-    const int input_rank = begin.getNumElements();
+    const uint64_t input_rank = begin.getNumElements();
     for (uint64_t i = 0; i < input_rank; i++) {
```
Comment (mihaimaruseac) [inline tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:1267]: Would rather change type of `i` in the loop.

File: tensorflow/compiler/mlir/lite/transforms/optimize.cc [hunk 11]
```diff
@@ -197,9 +197,9 @@ TypeAttr RescaleQtype(Type input, Attribute factor) {
 DenseElementsAttr GetShape(Value output_val) {
   auto output_type = output_val.getType().cast<RankedTensorType>();
   auto shape_vector = output_type.getShape();
-  std::vector<int32_t> shape(shape_vector.size());
-  for (int i = 0; i < shape_vector.size(); ++i) {
-    shape[i] = shape_vector[i];
+  std::vector<int32_t> shape;
+  for (auto shape_object : shape_vector) {
+    shape.push_back(shape_object);
```
Comment (mihaimaruseac) [inline tensorflow/compiler/mlir/lite/transforms/optimize.cc:202]: This changes memory allocation behavior as now the vector keeps being regrown. How about this?
```suggestion
  std::vector<int32_t> shape;
  shape.reserve(shape_vector.size());
  for (auto shape_object : shape_vector) {
    shape.push_back(shape_object);
```

File: tensorflow/compiler/mlir/lite/transforms/prepare_tf.cc [hunk 12]
```diff
@@ -495,7 +495,8 @@ struct ConvertTFStridedSlice : public RewritePattern {
         original_input_type.getShape();
     SmallVector<int64_t, 4> new_shape;
     int index = 0;
-    while (index < original_input_shape.size() || new_axis_mask) {
+    const int original_input_shape_size = original_input_shape.size();
+    while (index < original_input_shape_size || new_axis_mask) {
```
Comment (mihaimaruseac) [inline tensorflow/compiler/mlir/lite/transforms/prepare_tf.cc:499]:
```suggestion
    const int original_input_rank = original_input_shape.size();
    while (index < original_input_rank || new_axis_mask) {
```

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc (hunk 7 above)
Comment (tg-at-google) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:777]: updated.

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc (hunk 8 above)
Comment (tg-at-google) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:959]: Here auto seems to assign the objects being compared different types.
My attempt was to manually set the type of one of the two objects to resolve the warning.

I could use `int64_t` if `long int` is less appropriate.

Comment (mihaimaruseac) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:959]: `int64_t` is better

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc (hunk 7 above)
Comment (mihaimaruseac) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:777]: I don't see a name change

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by tg-at-google** [comment]
updated, 3.

**Review by mihaimaruseac — APPROVED** [review]
